Adding guestbook example with latest json format for v1beta3 api

sub-mod opened the pull request:
Support guestbook example json files to work with v3 api

sub-mod committed:
Adding guestbook example with namespace

derekwaynecarr commented:
Can I ask why this change is needed?

This basically means users will not be able to run this same example in another Kubernetes namespace.  The kubectl client will insert the namespace value from the current-context when omitted in the file, but when present, it errors if the context differs.

I don't this we should do this change.

sub-mod committed:
Removing namespace from json files

sub-mod commented:
I was just trying to give a working example of guestbook with latest k8-code and v3 api.I couldn't find one with latest format which worked on v3 api.
These json files have the latest json format .I agree hardcoding "default" can cause problems in other namespace.I gave "default" as the namespace just as an example so that the files were complete with all the data.But it was not really required since namespace field value can be fetched from the URI.
btw I  was using curl instead of kubectl.

If it is worth to have this example , I have updated the json files and removed the namespace field from all the files.Else please feel free to mark this PR as Closed with no change.

nikhiljindal commented:
@derekwaynecarr: Can I assigne this to you?

bgrant0607 commented:
We need a v1beta3 example. I'll take a look.

bgrant0607 commented on examples/guestbook/guestbook_ns/redis-master.json:
```diff
@@ -0,0 +1,32 @@
+{
+    "apiVersion": "v1beta3",
+    "kind": "Pod",
```
redis-master should be a 1-pod replication controller. See #4267

bgrant0607 commented on examples/guestbook/guestbook_ns/redis-master.json:
```diff
@@ -0,0 +1,32 @@
+                "ports": [
+                    {
+                        "containerPort": 6379,
+                        "hostPort": 6379,
```
Remove hostPort. #1487

bgrant0607 commented on examples/guestbook/guestbook_ns/redis-slave-controller.json:
```diff
@@ -0,0 +1,47 @@
+                        "ports": [
+                            {
+                                "containerPort": 6379,
+                                "hostPort": 6380
```
Remove hostPort.

bgrant0607 commented on examples/guestbook/guestbook_ns/redis-slave-service.json:
```diff
@@ -0,0 +1,19 @@
+        "selector": {
+            "name": "redisslave"
+        },
+        "sessionAffinity": "None"
```
Omit sessionAffinity. Shouldn't be necessary.

bgrant0607 replied:
I see that a number of defaults are specified, presumably as examples. I can understand the reason for that, but it makes our examples look unnecessarily complex, so I think we need to put fully populated objects elsewhere.

bgrant0607 commented on examples/guestbook/guestbook_ns/redis-slave-controller.json:
```diff
@@ -0,0 +1,47 @@
+    "metadata": {
+        "name": "redis-slave-controller",
+        "labels": {
+            "name": "redisslave",
```
We should hyphenate everything consistently.

bgrant0607 commented:
Please look at #1603 and the other issues/PRs cited.

Also, I'd like to take advantage of kubectl's multi-object capabilities. At least all objects relating to a particular microservice (e.g., frontend) should be in the same file.

bgrant0607 commented:
And we should add this to https://github.com/GoogleCloudPlatform/kubernetes/blob/master/examples/examples_test.go

bgrant0607 commented on examples/guestbook/guestbook_ns/frontend-controller.json:
```diff
@@ -0,0 +1,46 @@
+                "name": "frontend-controller",
+                "labels": {
+                    "name": "frontend",
+                    "uses": "redisslave,redis-master",
```
"uses" is not a pattern I want to endorse. Please remove it.

bgrant0607 commented on examples/guestbook/guestbook_ns/frontend-controller.json:
```diff
@@ -0,0 +1,46 @@
+        "name": "frontend-controller",
+        "labels": {
+            "name": "frontend",
+            "template": "frontend"
```
"template" isn't the right idea here. Besides, do we even need it? I'd maybe remove it. Alternatively, we could apply some of the typically labels we've discussed, such as those mentioned here: https://github.com/GoogleCloudPlatform/kubernetes/pull/3233#discussion_r22664461

bgrant0607 commented on examples/guestbook/guestbook_ns/frontend-controller.json:
```diff
@@ -0,0 +1,46 @@
+    "kind": "ReplicationController",
+    "apiVersion": "v1beta3",
+    "metadata": {
+        "name": "frontend-controller",
```
Please remove `-controller` from the name.

bgrant0607 commented on examples/guestbook/guestbook_ns/frontend-controller.json:
```diff
@@ -0,0 +1,46 @@
+        },
+        "template": {
+            "metadata": {
+                "name": "frontend-controller",
```
Please remove `-controller` from the name.

bgrant0607 commented on examples/guestbook/guestbook_ns/frontend-controller.json:
```diff
@@ -0,0 +1,46 @@
+                        "ports": [
+                            {
+                                "containerPort": 80,
+                                "hostPort": 8000
```
Please remove hostPort. Example instructions can show how to use the apiserver's proxy to access the server.

bgrant0607 commented on examples/guestbook/guestbook_ns/frontend-controller.json:
```diff
@@ -0,0 +1,46 @@
+                        ]
+                    }
+                ],
+                "restartPolicy": {
```
Please omit restartPolicy. It's not really configurable in a replication controller.

bgrant0607 commented on examples/guestbook/guestbook_ns/frontend_service.json:
```diff
@@ -0,0 +1,17 @@
+  "apiVersion": "v1beta3",
+  "metadata": {
+    "name": "frontend",
+    "labels": {},
```
By convention, we should apply the same labels to the service as those in its selector.

sub-mod committed:
new changes

sub-mod committed:
json files updated

sub-mod committed:
json files updated

sub-mod committed:
testcase updated

sub-mod committed:
add v3 json files for guesybook-go

bgrant0607 commented:
FYI, no email is sent when you push changes. You need to comment on the PR in order for someone to notice. Looking now.

bgrant0607 commented on examples/examples_test.go:
```diff
@@ -115,6 +115,14 @@ func TestExampleObjectSchemas(t *testing.T) {
 			"redis-master-service":   &api.Service{},
 			"redis-slave-service":    &api.Service{},
 		},
+		"../examples/guestbook/guestbook_ns": {
```
You have renamed the example.

sub-mod replied:
I have added it in a new folder called guestbook_ns.
If changes look good , I will merge with guestbook folder

sub-mod replied:
@bgrant0607 please let me know if it looks ok?

bgrant0607 replied:
You introduced the following directories:
```
examples/guestbook-go/guestbook-go_v3/
examples/guestbook/guestbook_ns/
```
They should be consistent. Also, the `guestbook_ns` example no longer uses namespace, and `guestbook` doesn't really need to be in the subdirectory names.

Both examples should be tested here.

bgrant0607 replied:
I suggest creating `v1beta3` subdirectories in each example, with a short README.md just stating that it's a v1beta3 version of the example, and explaining that v1beta3 needs to be enabled.

In a couple weeks we'll promote these to the top level and eliminate the v1beta1 examples.

sub-mod committed:
renaming folders

sub-mod committed:
updating testcases

sub-mod committed:
updating testcases

sub-mod commented:
@bgrant0607  I have made the changes, please let me know if it looks ok

sub-mod committed:
renaming folder to v1beta3 and updating testcases

sub-mod commented:
@bgrant0607  I have made the changes, please let me know if it looks ok

bgrant0607 commented on examples/guestbook-go/v1beta3/guestbook-service.json:
```diff
@@ -0,0 +1,19 @@
+        "port": 3000,
+        "containerPort": "http-server",
+        "protocol": "TCP",
+        "createExternalLoadBalancer": true,
```
We're not creating an external load balancer in the existing examples because it's not supported by all cloud providers.

sub-mod committed:
removing LoadBalancer

bgrant0607 commented on examples/guestbook-go/v1beta3/guestbook-controller.json:
```diff
@@ -0,0 +1,39 @@
+        },
+        "template": {
+            "metadata": {
+                "name": "guestbook",
```
I think this name is ignored:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/controller/replication_controller.go#L70

I'm guessing we're not validating that name isn't specified due to legacy v1beta1/2 support.

Can you confirm, @smarterclayton?

smarterclayton replied:
That is correct - argument for validating the versioned structs, not the internal.


> On Feb 23, 2015, at 5:33 PM, Brian Grant <notifications@github.com> wrote:
> 
> In examples/guestbook-go/v1beta3/guestbook-controller.json:
> 
> > +    "kind": "ReplicationController",
> > +    "apiVersion": "v1beta3",
> > +    "metadata": {
> > +        "name": "guestbook",
> > +        "labels": {
> > +            "name": "guestbook"
> > +        }
> > +    },
> > +    "spec": {
> > +        "replicas": 3,
> > +        "selector": {
> > +            "name": "guestbook"
> > +        },
> > +        "template": {
> > +            "metadata": {
> > +                "name": "guestbook",
> I think this name is ignored:
> https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/controller/replication_controller.go#L70
> 
> I'm guessing we're not validating that name isn't specified due to legacy v1beta1/2 support.
> 
> Can you confirm, @smarterclayton?
> 
> —
> Reply to this email directly or view it on GitHub.
>

bgrant0607 commented on examples/guestbook/v1beta3/frontend-controller.json:
```diff
@@ -0,0 +1,38 @@
+                        "image": "kubernetes/example-guestbook-php-redis",
+                        "ports": [
+                            {
+                                "containerPort": 80,
```
Nit: misaligned spacing

bgrant0607 commented on examples/guestbook/v1beta3/frontend-service.json:
```diff
@@ -0,0 +1,16 @@
+  "apiVersion": "v1beta3",
+  "metadata": {
+    "name": "frontend",
+    "labels": {"name": "frontend"},
```
In all other objects, you put the curly braces and labels on separate lines. Could you please follow that convention here, for consistency?

bgrant0607 commented on examples/guestbook/v1beta3/redis-master-service.json:
```diff
@@ -0,0 +1,18 @@
+    "metadata": {
+        "name": "redis-master",
+        "labels": {
+            "name": "redis-master"
```
Please make separate labels for "redis" and "master" as in the other example.

bgrant0607 replied:
As discussed, we can fix this when fixing (or removing) the original v1beta1 example.

sub-mod committed:
removing name from template

bgrant0607 commented on examples/guestbook-go/v1beta3/redis-master-controller.json:
```diff
@@ -0,0 +1,43 @@
+                        "image": "gurpartap/redis",
+                        "ports": [
+                            {
+                                "name": "redis-server",
```
Please use the port names everywhere, or eliminate them everywhere.

bgrant0607 replied:
If some of the port names are required, I'm ok with leaving them as is for now.

bgrant0607 commented on examples/guestbook/v1beta3/redis-slave-controller.json:
```diff
@@ -0,0 +1,39 @@
+    "metadata": {
+        "name": "redis-slave",
+        "labels": {
+            "name": "redis-slave"
```
Again, separate labels.

sub-mod replied:
kindly request to keep redis-master as is to keep back ward compatibility with initial brendanburns version

sub-mod committed:
format all json files

bgrant0607 commented:
LGTM. Will merge on green. Thanks.

bgrant0607 commented:
Shippable failures were 2 different flakes. Merging.

fgrzadkowski commented on examples/guestbook-go/v1beta3/README.md:
```diff
@@ -0,0 +1,14 @@
+The web front end interacts with the redis master via javascript redis API calls.
+
+The v1beta3 API is not enabled by default. The kube-apiserver process needs to run with the --runtime_config=api/v1beta3 argument. Use the following command to enable it:
+$sudo sed -i 's|KUBE_API_ARGS="|KUBE_API_ARGS="--runtime_config=api/v1beta3 |' /etc/kubernetes/apiserver
```
This doesn't work for me. Where should I run it? On master? On my local machine?

sub-mod replied:
add this in /etc/kubernetes/apiserver where apiserver is running
KUBE_API_ARGS="--runtime_config=api/v1beta3"
